[MNT] resolve duplication in KNeighborsClassifier and KNeighborsRegressor

Z-Fran opened the pull request:
#### Reference Issues/PRs
https://github.com/sktime/sktime/pull/6217#pullrequestreview-1969913623

#### What does this implement/fix?
create a new class `KNeighborsSklearnAdapter` to resolve duplication in KNeighborsClassifier and KNeighborsRegressor

#### What should a reviewer concentrate their feedback on?
1. whether the path of the new file is appropriate
2. whether the name of the new class is appropriate

fkiraly reviewed (changes requested):
Interesting, thanks! I think this is a robust refactor, as you say we need to think about the right target location.

Might I suggest to split the adapter in two, and move the parts as follows:

* the distance adapter logic should to in `dists_kernels/base/adapters/_sklearn`
* the common `_fit`, `predict` features should go in `base._panel.knn` (and submodules thereof)

Z-Fran commented:
I have moved the distance adapter logic to `dists_kernels/base/adapters/_sklearn` and moved `_fit`, `predict` features to `base._panel.knn`. But I don't know how to organize `base._panel.knn`. well. I need some of your suggestions. @fkiraly

fkiraly commented on sktime/base/_panel/knn.py:
```diff
@@ -0,0 +1,103 @@
+__all__ = ["BaseKnnTimeSeriesEstimator"]
+
+
+class BaseKnnTimeSeriesEstimator:
```
make this a private class, start it with underscore, or it might be picked up by the tests.

fkiraly commented on sktime/dists_kernels/base/adapters/_sklearn.py:
```diff
@@ -0,0 +1,86 @@
+from sktime.datatypes import convert
+
+
+class _SklearnKnnAdapter:
```
not related to distance, so I would call it `DistAdapter`

fkiraly replied:
it's also a mixin, so how about `_SklearnDistMixin`

fkiraly commented on sktime/dists_kernels/base/adapters/_sklearn.py:
```diff
@@ -0,0 +1,86 @@
+class _SklearnKnnAdapter:
+    """Base adapter mixin for sklearn distances of KNeighbors."""
+
+    _tags = {
```
not an estimator, so no tags needed

fkiraly commented on sktime/base/_panel/knn.py:
```diff
@@ -0,0 +1,103 @@
+class BaseKnnTimeSeriesEstimator:
+    """Base KNeighbors Time Series Estimator."""
+
+    _tags = {
```
mixin, so no tags needed

Z-Fran commented:
Hi @fkiraly ,  I have finished your suggestions.

fkiraly commented on sktime/classification/distance_based/_time_series_neighbors.py:
```diff
@@ -230,90 +232,14 @@ def _distance(self, X, X2=None):
         # sktime distance classes are Panel x Panel -> numpy2D
         # and the numba distances are numpy3D x numpy3D -> numpy2D
         # so we need to wrap the sktime distances
+        self._X = X
```
why are we doing this?

fkiraly reviewed (changes requested):
looks good!

One minor question, why are we writing `self._X`? That does not look like a refactor, but a change in logic.

Z-Fran commented:
> looks good!
> 
> One minor question, why are we writing `self._X`? That does not look like a refactor, but a change in logic.

Oh, it's related to [#6515 ](https://github.com/sktime/sktime/issues/6515#issuecomment-2144002401). I wrote it when debugging. I've deleted it.

fkiraly commented:
Ah, I see! I wuld suggest, let's try to fix that (if at all) in another PR, and add tests - this PR should be a pure refactor, i.e., not changing logic.

fkiraly reviewed (approved):
Looks good to me now!

Z-Fran commented:
Hi, @fkiraly , could it be merged?

fkiraly commented:
Yes, that´s the plan. Just wanted to give other core devs some time to react. Best practice is 5 working days.
